### PR TITLE
fix(client-go): paginate over lists

### LINKED ISSUE #3911: bug(client): deis apps only shows 100 apps
The client is not using the pagination to return all the apps

### PR DESCRIPTION
**Pull request by Joshua-Anderson**
Right now, if you have over 100 apps, deis only lists the first 100, now all of the apps are displayed.

Fixes #3911

### CODE REVIEW

**Comment by bacongobbler** [comment]
coed LGTM

**Comment by krancour** [comment]
Hypothetically, if an app had 1,000 releases (not at all unrealistic in the era of continuous deployments), this means the client makes 10 HTTP requests to stitch together the full list of 1,000 releases?  This doesn't seem ideal and seems as if it may defeat the purpose of paginating those responses in the first place.

What would be your thoughts on allowing users to interactively scroll through the results, triggering new page requests only when they scroll past the end of what's already been received?  (Similar to what happens, for instance, with large result sets in `psql`.)  I know this is easier said than done, but does it make more sense for the users?

**Comment by carmstrong** [comment]
> What would be your thoughts on allowing users to interactively scroll through the results

I think that would be difficult for API clients. I think it's up the the client what it wants - it should be able to specify a "results per page". So an API client could request 10,000 results per page to get everything in one request.

**Comment by krancour** [comment]
> So an API client could request 10,000 results per page to get everything in one request.

But that doesn't answer the question of how _this_ client should deal with large result sets.  Regardless of whether it makes 1 request to get 1,000 results, or 10 requests to get 1,000 results, the user probably doesn't want to _see_ 1,000 results-- which means in either case, we pulled down too much stuff.

**Comment by carmstrong** [comment]
Maybe for this case, cap it to 100 or something by default, and display a message like "More than 100 apps exist." That cap could be toggled by a parameter to `deis ps`.

**Comment by krancour** [comment]
@carmstrong :+1: same goes for `deis apps`, `deis releases`, etc.

**Comment by krancour** [comment]
```
Options:
  -l --limit=<limit>
    the maximum number of results to display; defaults to 1000
```

**Comment by carmstrong** [comment]
:+1:

**Comment by Joshua-Anderson** [comment]
I'm now using the limit based approach to paginate!

File: client-go/controller/models/apps/apps_test.go [hunk 1]
```diff
@@ -291,7 +291,7 @@ func TestAppsList(t *testing.T) {
 
 	client := client.Client{HTTPClient: httpClient, ControllerURL: *u, Token: "abc"}
 
-	actual, err := List(&client)
+	actual, err := List(&client, 100)
```
Comment (bacongobbler) [inline client-go/controller/models/apps/apps_test.go]: I feel like this [magic number](https://en.wikipedia.org/wiki/Magic_number_%28programming%29) should be part of the client config with a default of 100. That way users can tune this number by modifying ~/.deis/client.json

Comment (Joshua-Anderson) [inline client-go/controller/models/apps/apps_test.go]: Good Idea! I implemented it. In this particular location these are just the tests and I just gave a random value as it doesn't affect the response.

**Comment by krancour** [comment]
I like the implementation overall.  My only outstanding concern is that with the limit-based approach, the frequent use of "paginate" or "paginated" in the code (e.g. in function names) strongly hints at the existence of a feature that is _not_ there-- the ability to actually "turn the page" so-to-speak (which would require handling an offset in addition to the limit).  I don't think that feature needs to be added or anything, but adapting the function names to no longer imply responses are "pageable" might be a good thing.

**Comment by Joshua-Anderson** [comment]
@krancour I agree. I renamed the functions to be less ambiguous.

**Comment by krancour** [comment]
@Joshua-Anderson on the controller end, would `limit` be a better query param than `page_size`?  Notion of pages still appears in some comments, too.

**Comment by Joshua-Anderson** [comment]
@krancour The reason I went with `?page_size` is because it is a standard and used by almost everything that uses pagination. Most of the time it is used like `\paginate?page=2&page_size=80`.

> Notion of pages still appears in some comments, too.

If so, it's accidental and please leave a comment at the location so I can fix it.

**Comment by krancour** [comment]
> The reason I went with ?page_size is because it is a standard and used by almost everything that uses pagination.

So is the idea here that the controller will allow for pagination, even though the client really only supports retrieval of one page of specifiable size?

If so, I think reference to "page" or "paginated" make sense on the controller side, and I'll look for just the errant references to "page" or "paginated" on the CLI side.

File: client-go/controller/client/client.go [hunk 2]
```diff
@@ -26,13 +26,20 @@ type Client struct {
 
 	// Username is the name of the user performing requests against the API.
 	Username string
+
+	// ResponseLimit is the number of results to return on a paginated requests.
```
Comment (krancour) [inline client-go/controller/client/client.go]: "paginated" might confuse here.  The controller supports it, but the CLI can only ever fetch one "page," even though its size is specifiable.  Same comment applies on line 34.

**Comment by Joshua-Anderson** [comment]
@krancour Done!

**Comment by krancour** [comment]
LGTM

File: client-go/cmd/apps.go [hunk 3]
```diff
@@ -55,14 +55,18 @@ func AppCreate(id string, buildpack string, remote string, noRemote bool) error
...
-func AppsList() error {
+func AppsList(results int) error {
 	c, err := client.New()
 
 	if err != nil {
 		return err
 	}
 
-	apps, err := apps.List(c)
+	if results == -1 {
```
Comment (mboersma) [inline client-go/cmd/apps.go]: Should this magic number `-1` be a constant instead, like `DEFAULT_LIMIT`?

File: client-go/controller/client/client_test.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
 	"testing"
 )
 
-const sFile string = `{"username":"t","ssl_verify":false,"controller":"http://d.t","token":"a"}`
+const sFile string = `{"username":"t","ssl_verify":false,"controller":"http://d.t","token":"a", "response_limit": 50}`
```
Comment (mboersma) [inline client-go/controller/client/client_test.go]: Minor, but could you omit the spaces in this JSON string? It just reads strangely to have inconsistent spacing here.

**Comment by mboersma** [comment]
Needs a rebase against master now. :repeat_one:

**Comment by Joshua-Anderson** [comment]
@mboersma Rebased!

**Comment by carmstrong** [comment]
@bacongobbler @mboersma Would you mind re-reviewing this? I don't feel comfortable enough with the code.

**Comment by mboersma** [comment]
LGTM.

**Comment by krancour** [comment]
I've been banging pretty hard on the k8s scheduler.  Something I can easily do now that wasn't previously realistic is scale an app to hundreds of containers very quickly.  In one case, I scaled to 200.  `deis ps` only shows me the first hundred (and I know that's because 100 is the default limit we discussed above), but I see no indication that I'm viewing a partial list.  Would it possibly make sense, for certain queries, to return and display a summary of the full result set?

**Comment by Joshua-Anderson** [comment]
@krancour Could you give some example output for what you would like to see?

**Comment by krancour** [comment]
So, immediately after scaling, I see this:

```
[kent@mbp example-go]$ deis scale web=200
Scaling processes... but first, coffee!
done in 94s
=== jicama-jamboree Processes

--- web: 
web.1 up (v2)
... omitted for brevity...
web.100 up (v2)
```

This shows me 100 processes, even though I just scaled to 200.  For anyone who doesn't know about limits, this looks very odd.  I see the same result as above when using `deis ps`.

It seems a lot of clarity could be added by having the output resemble something like this:

```
[kent@mbp example-go]$ deis scale web=200
Scaling processes... but first, coffee!
done in 94s
=== jicama-jamboree Processes (100 of 200)

--- web: 
web.1 up (v2)
... omitted for brevity...
web.100 up (v2)
```

That sort of `(m of n)` info could be generally useful and insightful for other queries as well.  e.g. `deis apps` or `deis users`.  In the case of `deis ps`, it would actually be possible to also summarize the status of the processes with an additional line that shows the count in each status-- how many are up, how many are down, etc., but that seems like it's probably not totally necessary.  If we're at least making sure users _know_ they're seeing only a fragment of the results, they can at least choose to query again with a higher limit and maybe `grep` it themselves for whatever it is they are looking for.

**Comment by Joshua-Anderson** [comment]
@krancour  :+1: Like the idea.

**Comment by Joshua-Anderson** [comment]
@krancour I added the `(100 of 200)` message to every limited request (expect certs, because it has a different layout.

**Comment by krancour** [comment]
:+1: I'll re-review as soon as I've got a chance, but let me preemptively say-- nice work on this.

**Comment by bacongobbler** [comment]
manually tested, works as intended:

```
><> deis ps:list -a go
=== go Processes (100 of 101)
```

LGTM!

**Comment by krancour** [comment]
LGTM

**Pull request merged**